src/workerd/api/sql.c++

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+#include "sql.h"
+#include <stdio.h>
+
+namespace workerd::api {
+
+SqlResult::SqlResult(SqliteDatabase::Query& query) {
+  uint columnCount = query.columnCount();
+
+  while (!query.isDone()) {
+    kj::Vector<jsg::Dict<SqliteDatabase::Query::ValueOwned>::Field> fields;
+    for (uint i = 0; i < columnCount; ++i) {
+      fields.add(jsg::Dict<SqliteDatabase::Query::ValueOwned>::Field{
+        .name = kj::heapString(query.getColumnName(i)),
+        .value = query.getValueOwned(i)
+      });
+    }
+    results.add(jsg::Dict<SqliteDatabase::Query::ValueOwned>{.fields = fields.releaseAsArray()});
+    query.nextRow();
+  }
+}
+
+kj::ArrayPtr<jsg::Dict<SqliteDatabase::Query::ValueOwned>> SqlResult::getResults(jsg::Lock& js) {
+  return results;
+}
+
+SqlPreparedStatement::SqlPreparedStatement(jsg::Ref<SqlDatabase>&& sqlDb, SqliteDatabase::Statement&& statement):
+  sqlDatabase(kj::mv(sqlDb)),
+  statement(kj::mv(statement)) {
+}
+
+void fillBindValues(
+  kj::Vector<SqliteDatabase::Query::ValuePtr>& bindValues, kj::Maybe<kj::Array<ValueBind>>& bindValuesOptional) {
+  KJ_IF_MAYBE(values, bindValuesOptional) {
+    for (auto& value : *values) {
+      KJ_SWITCH_ONEOF(value) {
+        KJ_CASE_ONEOF(val, kj::Array<const byte>) {
+          bindValues.add(val.asPtr());
+        }
+        KJ_CASE_ONEOF(val, kj::String) {
+          bindValues.add(val.asPtr());
+        }
+        KJ_CASE_ONEOF(val, double) {
+          bindValues.add(val);
+        }
+      }
+    }
+  }
+}
+
+jsg::Ref<SqlResult> SqlPreparedStatement::run(kj::Maybe<kj::Array<ValueBind>> bindValuesOptional) {
+  kj::Vector<SqliteDatabase::Query::ValuePtr> bindValues;
+  fillBindValues(bindValues, bindValuesOptional);
+
+  kj::ArrayPtr<const SqliteDatabase::Query::ValuePtr> boundValues = bindValues.asPtr();
+  SqliteDatabase::Query query(sqlDatabase->sqlite, statement, boundValues);
+  return jsg::alloc<SqlResult>(query);
+}
+
+SqlDatabase::SqlDatabase():
+  dir(kj::newInMemoryDirectory(kj::nullClock())),
+  vfs(*dir),
+  sqlite(vfs, kj::Path({"root"}), kj::WriteMode::CREATE | kj::WriteMode::MODIFY) {
+}
+
+jsg::Ref<SqlDatabase> SqlDatabase::constructor(jsg::Lock& js, CompatibilityFlags::Reader flags) {
+  return jsg::alloc<SqlDatabase>();
+}
+
+jsg::Ref<SqlResult> SqlDatabase::exec(jsg::Lock& js, kj::String querySql, kj::Maybe<kj::Array<ValueBind>> bindValuesOptional) {
```

Comment:
At present there's no protection against SQL code using dangerous pragmas. I think it's important that we lock this down early before we expose this API to JS. We can use SQLite's authorizer callbacks:

https://www.sqlite.org/c3ref/set_authorizer.html

In addition to blocking weird pragmas and such, we need to block access to any table whose name starts with `_cf_`, in order to hide the DO KV table. (We want people to query that only through the explicit API because it contains V8-serialized values and we don't really want to expose the serialization.)